File: cards_group_914/cards_group.go
package cards_group_914

import "fmt"

func hasGroupsSizeX(deck []int) bool {
	if len(deck) == 1 {
		return false
	}

	m := make(map[int]int)
	for _, v := range deck {
		m[v] = m[v] + 1
	}
	g := m[0] //最大公约数
	for i := 1; i < len(deck); i++ {
		g = gcd(g, m[i])
	}
	if g == 1 {
		return false
	}
	return len(deck)%g == 0
}

func gcd(a, b int) int {
	if b == 0 {
		return a
	}
	return gcd(b, a%b)
}

func main() {
	aaa := gcd(6, 12)
	fmt.Println(aaa)
	
}

File: data_2020_04/cards_group_914/cards_group.py
# -*- coding: utf-8 -*-
import math

class Solution:
    def hasGroupsSizeX(self, deck) -> bool:
        D = {}
        for x in deck:
            D[x] = D.setdefault(x, 0) + 1
        a = []
        for v in D.values():
            a.append(v)

        print(D, a)
        gcd = a[0]
        for i in range(1, len(a)):
            gcd = math.gcd(gcd, a[i])
        if gcd == 1:
            return False
        return True

"""
输入：[1,2,3,4,4,3,2,1]
输出：true
解释：可行的分组是 [1,1]，[2,2]，[3,3]，[4,4]
"""